### PR TITLE
Make checkstyle work with toolchains and worker api

### PR DESCRIPTION
**Pull request by breskeby**
This Pull request makes the checkstyle task use the worker api. It will use the current jdk or java toolchain defined or you can declare a specific toolchain per Checkstyle task.  

The checkstyle task will always use the worker api now.

This is based on https://github.com/gradle/gradle/pull/18312 

It addresses the review feedback of @big-guy made on This is based on https://github.com/gradle/gradle/pull/18312 

### Context
Changing the checkstyle tasks to rely on the worker api allows parallel execution on checkstyle tasks within a project making significant performance improvements in projects with multiple larger sourceSets.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by breskeby** [comment]
I don't see how failing `sanityCheck` is related to these changes

File: subprojects/core/src/main/java/org/gradle/testfixtures/internal/ProjectBuilderImpl.java [hunk 1]
```diff
@@ -107,6 +107,8 @@ public ProjectInternal createProject(String name, File inputProjectDir, File gra
         File userHomeDir = gradleUserHomeDir == null ? new File(projectDir, "userHome") : FileUtils.canonicalize(gradleUserHomeDir);
         StartParameterInternal startParameter = new StartParameterInternal();
         startParameter.setGradleUserHomeDir(userHomeDir);
+        startParameter.useEmptySettings();
+        startParameter.setCurrentDir(inputProjectDir);
```
Comment (JLLeitschuh) [inline subprojects/core/src/main/java/org/gradle/testfixtures/internal/ProjectBuilderImpl.java:111]: Why are these changes also needed as a part of this larger change?

Comment (breskeby) [inline subprojects/core/src/main/java/org/gradle/testfixtures/internal/ProjectBuilderImpl.java:111]: added to get FileFactory working correctly with ProjectBuilder

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/AntWorkAction.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            Class<?> buildListenerClass = cl.loadClass("org.apache.tools.ant.BuildListener");
+            Method addBuildListener = projectClass.getDeclaredMethod("addBuildListener", buildListenerClass);
+            Method removeBuildListener = projectClass.getDeclaredMethod("removeBuildListener", buildListenerClass);
+            Method getBuildListeners = projectClass.getDeclaredMethod("getBuildListeners");
+            Vector<?> listeners = (Vector<?>) getBuildListeners.invoke(project);
+            removeBuildListener.invoke(project, listeners.get(0));
+            addBuildListener.invoke(project, antLogger);
+        } catch (Exception ex) {
+            throw UncheckedException.throwAsUncheckedException(ex);
+        }
```
Comment (JLLeitschuh) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/AntWorkAction.java:71]: This exception catching looks to be overly broad IMHO. Also, returning a more useful exception would probably be valuable here as well.

**Review by JLLeitschuh — CHANGES_REQUESTED** [review]
To be merged, this will need release notes added as well.

File: subprojects/core/src/main/java/org/gradle/testfixtures/internal/ProjectBuilderImpl.java (hunk 1 above)
Comment (breskeby) [reply subprojects/core/src/main/java/org/gradle/testfixtures/internal/ProjectBuilderImpl.java:111]: The checkstyle plugin uses the FileFactory service. Without these changes the invoked service in the Project mockup reports the wrong folder and not the temp folder project root as expected

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/AntWorkAction.java (hunk 2 above)
Comment (breskeby) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/AntWorkAction.java]: I made them a little bit more explaining. In theory they _should_ not happen though just wrapping reflection exceptions here on the antbuilder

**Comment by breskeby** [comment]
> To be merged, this will need release notes added as well.

I've added  a little paragraph to the release notes

**Comment by breskeby** [comment]
@JLLeitschuh addressed your review feedback

**Comment by ljacomet** [comment]
Thanks a lot for moving forward with this. However the window is closed for 7.4 so this will be considered for 7.5.
The Gradle team will follow up on this early next year.

**Comment by breskeby** [comment]
@ljacomet Can we consider this for 7.5? It brings a huge improvement on build time for our multi project builds with projects containing multiple sourceSets

**Comment by asodja** [comment]
I tested this on `gradle/gradle` and I did not see much improvement or regression. There is more `gc` time spent with new version, but other than that it looks good. And we want to support toolchains in the checkstyle task, so we want to go with this forward.
Benchmarks: [tests.zip](https://github.com/gradle/gradle/files/8152241/tests.zip)

@breskeby what kind of improvements did you see?

**Comment by breskeby** [comment]
In our multi project build we have plenty of projects with more than 2 sourceSets (usually additional test sourceSets) for which we also run checkstyle. With this change we see significant improvements because checkstyle tasks within one subproject can now run in parallel. 

Here's an example of running the precommit check (includes those checkstyle tasks) against our biggest subproject before and after this change:

Before:
Screenshot 2022-02-28 at 11 00 46

After:
Screenshot 2022-02-28 at 11 00 35

**Comment by asodja** [comment]
@breskeby Thanks for that graphs and info. That is a nice improvement, awesome.

I will update this branch and run tests again.

And thank you for pushing that forward.

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@breskeby can you sign-off your commits? DCO is complaining with:
```
Summary
Commit sha: [94c9e3f](https://github.com/gradle/gradle/pull/19361/commits/94c9e3f670a9893a3a5b57a0c71acabca5d90b7d), Author: Rene Groeschke, Committer: Rene Groeschke; The sign-off is missing.
Commit sha: [c8ea02e](https://github.com/gradle/gradle/pull/19361/commits/c8ea02e4e51433740592a198c1af3b550d1b21a1), Author: Rene Groeschke, Committer: Rene Groeschke; The sign-off is missing.
Commit sha: [140f69b](https://github.com/gradle/gradle/pull/19361/commits/140f69b378576027953721ef58a2224ca79ff473), Author: Rene Groeschke, Committer: Rene Groeschke; The sign-off is missing.
Commit sha: [a2ee9b4](https://github.com/gradle/gradle/pull/19361/commits/a2ee9b448197cf9cbad525cfff7d0c34b6101f02), Author: Rene Groeschke, Committer: Rene Groeschke; The sign-off is missing.
Commit sha: [d4a42dc](https://github.com/gradle/gradle/pull/19361/commits/d4a42dcd5a443bb7468ce4feaaa49a5f48bb53bc), Author: Rene Groeschke, Committer: Rene Groeschke; The sign-off is missing.
```
If you agree I can also rebase them so I am co-author of these commits. I think in that case DCO will be happy.

**Comment by asodja** [comment]
Since I already did some work on this branch, could you reset a branch back to your last commit `d4a42dc` and sign-off just your added commits. And then I will do the rest to get back to current state.

**Comment by breskeby** [comment]
@asodja pushed my signed off commits to https://github.com/breskeby/gradle/tree/checkstyle-wapi-signedoff sorry for the inconvenience. should have done that in the first place

**Comment by asodja** [comment]
Awesome, thank you!

**Comment by asodja** [comment]
I will close this and continue in https://github.com/gradle/gradle/pull/20069.

**Pull request closed without merging**